### PR TITLE
Rewrite schemas and implement custom types

### PR DESCRIPTION
**Pull request by georgolden**
- [x] tests and linter show no problems (`npm t`)
- [x] tests are added/updated for bug fixes and new features
- [x] code is properly formatted (`npm run fmt`)
- [ ] description of changes is added in CHANGELOG.md
- [ ] update .d.ts typings

Need help:
 - or ideas for Preprocessor - improve code quality
 - overall - better naming
 - Model change reorder to remove hardcoded projection

Non breaking:
 - elegant syntax/format for custom types and internal types
 - custom types support
 - preprocessor to reduce Schema preprocess complexity
 - any kind of nested arrays, array of references https://github.com/metarhia/metaschema/issues/378
 - nested schemas with Schema instances support
 - any level of complex nested types support
 - custom validate for field
 - changed validate for schema, simplified for user
 - all user's validate functions support 4 types of syntax: boolean return, throw Error, error message as string return and array of error messages
 - reserved words permitted in schema if kind provided
 - function fields stored in schema
 - shorthand required key for collections support `{ 'array?': 'string' }`
 - many relation now checks in runtime
 - model now loads projections at the end, to fix bug
 - schema kinds moved to separate file and kinds now have logic to remove hardcoded projection from Schema
 - ts interfaces from schema now have relation ids as well
 - deps update metautil
 - nested object support https://github.com/metarhia/metaschema/issues/395
 - syntax for validate function changed

Breaking:
 - Model no more require metavm for browser compatibility: need `impress` update
 - loadModel moved to loader: need `impress` update
 - Model no more writes d.ts file to disk moved to loader as well: need `metasql` update
 - syntax for pg types changed a bit, no more not-working types: need `metasql` update
 - not supporting custom Schema kinds: need to not use custom kinds in schemas folder: `application/schemas/<YourSchema.js>`

### CODE REVIEW

File: lib/prototypes/reference.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  format(def) {
+    const reference = def.one || def.many;
+    const relation = def.many ? 'many-to-one' : 'one-to-many';
+    this.type = reference;
+    this.references.add(reference);
+    this.relations.add({ to: reference, type: relation });
+  },
+};
+
+module.exports = { reference, many: reference, one: reference };
```
Comment (tshemsedinov) [inline lib/prototypes/reference.js:26]: Why do we need to export same object 3 times?

File: lib/prototypes/collections.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+const set = {
+  ...array,
+  type: 'set',
+  isInstance(value) {
+    return Reflect.getPrototypeOf(value) === Set.prototype;
+  },
+};
+
+module.exports = { object, map, array, set };
```
Comment (tshemsedinov) [inline lib/prototypes/collections.js:91]: Maybe it's better to hold each type in a separate file like `({ /* JSON5 metadata */ })` and load it with metavm

File: test/database.js [hunk 3]
```diff
@@ -3,9 +3,9 @@
 const metatests = require('metatests');
 const { Schema } = require('..');
 
-metatests.test('Schema: database', (test) => {
+metatests.test('Database: schema Registry', (test) => {
   const raw = {
-    Registry: {},
+    Registry: { scope: 'application', store: 'persistent', allow: 'write' },
```
Comment (tshemsedinov) [inline test/database.js:8]: Those fields are expected to be default

File: lib/prototypes/reference.js (hunk 1 above)
Comment (georgolden) [reply lib/prototypes/reference.js:26]: It will be included in Schema and checked as well. In previous version no fields like this were checked

Comment (georgolden) [reply lib/prototypes/reference.js]: I made improvement. Now moved to prototypes

**Comment by georgolden** [comment]
@tshemsedinov https://github.com/metarhia/metaschema/issues/395 Done

File: lib/schema.js [hunk 4]
```diff
@@ -1,214 +1,111 @@
...
+const RESERVED = ['validate', 'parse', 'serialize', 'format'];
 
 class Schema {
-  constructor(name, raw, namespaces) {
-    const short = shorthand(raw) || raw.type;
-    const defs = short ? { value: raw } : raw;
-    this.name = name;
-    this.kind = short ? 'scalar' : 'struct';
+  constructor(name, raw, namespaces = []) {
+    Schema.KIND = KIND;
```
Comment (lundibundi) [inline lib/schema.js:22]: Why is this here and not along with other static properties?

File: lib/schema.js [hunk 5]
```diff
@@ -337,7 +217,6 @@ class Schema {
   }
 }
 
-Schema.KIND = KIND;
 Schema.KIND_STORED = KIND_STORED;
```
Comment (lundibundi) [inline lib/schema.js:220]: Minimum supported Node.js version by metaschema is 12.9 so we can just use `static` properties. https://node.green/#ES2022-features-static-class-fields-public-static-class-fields

File: lib/util.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    }
+  },
+};
+
+const harmonize = (result) => {
+  if (typeof result === 'boolean') {
+    return result ? '' : ['Validation error'];
+  }
+  if (result) return Array.isArray(result) ? result : [result];
+};
```
Comment (lundibundi) [inline lib/util.js:41]:
```suggestion
const harmonizeErrors = (result) => {
  if (typeof result === 'boolean') {
    return result ? null : ['Validation error'];
  }
  if (result) return Array.isArray(result) ? result : [result];
  return null;
};
```
wdyt?

File: lib/schema.js [hunk 7]
```diff
@@ -1,214 +1,111 @@
...
-        const parent = this.findReference(this.parent);
-        for (const key of metadata.fields) {
-          defs[key] = parent.fields[key];
-          keys.push(key);
-        }
-      }
+  preprocess(raw, prep) {
+    const source = prep.parse(raw);
+    if (isType(source)) {
+      this.fields = source;
```
Comment (lundibundi) [inline lib/schema.js:56]: Does this mean that `fields` is not longer just an object with fields? This would probably break some things, why do we need this?

File: lib/prototypes/abstract.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  checks: {},
+  check(value, path, root) {
+    try {
+      const typeErr = harmonize(this.checkType(value, path, root));
+      if (typeErr) return typeErr;
+      if (this.validate) {
+        const validErr = harmonize(this.validate(value, path, root));
+        if (validErr) return validErr;
+      }
+      for (const [name, subCheck] of Object.entries(this.checks)) {
```
Comment (lundibundi) [inline lib/prototypes/abstract.js:16]: Why not have `checks` as Map or even an Array then?

File: lib/prototypes/abstract.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        if (this[name]) subCheck(value, path, this);
```
Comment (lundibundi) [inline lib/prototypes/abstract.js:17]: What if one of the `subCheck` wants to return an error, shouldn't we collect these into an array and return it?

File: lib/prototypes/abstract.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    const { formatters } = this;
+    for (const key of Object.keys(def)) {
+      if (key === 'type' || key === this.type) continue;
+      if (formatters[key]) this[key] = formatters[key](def[key]);
+      else this[key] = def[key];
+    }
+    this.construct(def, preprocessor);
+    const { type } = this;
+    this.type = type;
+    this.references.add(type);
```
Comment (lundibundi) [inline lib/prototypes/abstract.js:48]:
```suggestion
    this.references.add(this.type);
```
?

File: lib/prototypes/collections.js [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  entries(value) {
+    return Object.entries(value);
+  },
+};
+
+const map = {
+  ...object,
+  type: 'map',
+  isInstance(value) {
+    return Reflect.getPrototypeOf(value) === Map.prototype;
```
Comment (lundibundi) [inline lib/prototypes/collections.js:53]: This would probably not work in metavm contexts and same goes for other such checks, since they prototypes from different sandboes will probably not be reference-equal. Have you verifies this case? (`schema` created in one sandbox and `value` created in another)

File: lib/prototypes/collections.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    const source = def.type ? def.value : def[type];
+    const child = prep.parse(source);
+    if (isType(child)) {
+      this.value = child;
+      this.metadataFrom(child);
+    } else {
+      const schema = new prep.types.schema(child.defs, prep);
+      this.references.add('schema');
+      this.metadataFrom(schema);
+      this.value = schema;
```
Comment (lundibundi) [inline lib/prototypes/collections.js:84]: why are we storing schema under `value` field?

File: lib/schema.js [hunk 13]
```diff
@@ -233,15 +130,24 @@ class Schema {
...
+    const type = DEFAULT[name];
+    if (type) return type;
     return null;
   }
 
+  getTypes() {
+    let types = {};
+    for (const ns of this.namespaces) {
+      if (ns.types) types = { ...types, ...ns.types };
+    }
```
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/schema.js:146]: will be referenced in a commit. Thx a lot

File: lib/schema.js [hunk 14]
```diff
@@ -233,15 +130,24 @@ class Schema {
...
+    if (Object.keys(types).length === 0) types = DEFAULT;
+    return types;
```
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/schema.js:148]: will be referenced in a commit. Thx a lot

File: lib/schema.js (hunk 5 above)
Comment (georgolden) [reply lib/schema.js:220]: The reason for this is outdated `"eslint-config-metarhia"`. It depends on old eslint, so I got eslint error, which can not be disabled easily without update. I see no big issue with this approach. Syntax is OK. But we need to update our linter to support new ES features

File: lib/util.js (hunk 6 above)
Comment (georgolden) [reply lib/util.js:41]:
```js
const harmonize = (error) => {
  if (typeof error === 'boolean') {
    return error ? null : ['Validation error'];
  }
  if (error) return Array.isArray(error) ? error : [error];
  return null;
};
```
wdyt?

File: lib/schema.js (hunk 7 above)
Comment (georgolden) [reply lib/schema.js:56]: In case if schema is a single type, we don't need to keep useless field `{ value: { type: 'string' } }` we can store any type shorthand directly in fields and check by constructor name with `isType`

Comment (georgolden) [reply lib/schema.js:56]: Logs look like this:
```js
fields: Type {
    references: Set(1) { 'string' },
    relations: Set(0) {},
    required: true,
    type: 'string'
  },
```

File: lib/util.js (hunk 6 above)
Comment (georgolden) [reply lib/util.js:41]: Merged your changes and kept short name

File: lib/prototypes/abstract.js (hunk 8 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:16]: It never will be large, so no reason to use map. But I will try an array

File: lib/prototypes/abstract.js (hunk 9 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:17]: I did not use harmonize for subCheck. And I do not expect to run all checks, so I will add this option, cause we have similar behavior in Schema. Thx

File: lib/prototypes/abstract.js (hunk 10 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:48]: this needed for type to be visible in logs, otherwise type will be always in prototype and will be not visible in logs. And for reference it is a difference. Diff:
```js
apartment: Type {
      references: [Set],
      relations: Set(0) {},
      required: true,
      type: 'string'
    },
```
```js
persons: Type {
      references: [Set],
      relations: [Set],
      many: 'Person',
      required: true,
      type: 'Person'
    }
```

Comment (georgolden) [reply lib/prototypes/abstract.js:48]: Without this lines:
```js
apartment: Type {
      references: [Set],
      relations: Set(0) {},
      required: true
    },
```
```js
persons: Type {
      references: [Set],
      relations: [Set],
      many: 'Person',
      required: true,
      type: 'Person'
    }
```

File: lib/prototypes/collections.js (hunk 11 above)
Comment (georgolden) [reply lib/prototypes/collections.js:53]: This is from old version. Replaced with checking constructor name

File: lib/prototypes/collections.js (hunk 12 above)
Comment (georgolden) [reply lib/prototypes/collections.js:84]: schema is only invariant, that keeps exist. So it requires quite complex explanation to explain, why we can't avoid schema as an invariant. In short: array defenition has a reserved field called value, to describe type of each element in array.
```js
{ type: 'array', value: 'string' }
{ type: 'array', value: { name: 'string' } }
```

Comment (lundibundi) [reply lib/prototypes/collections.js:84]: So this is for `array` only, right?

Comment (georgolden) [reply lib/prototypes/collections.js:84]: for array and set - `{ value: Type/Schema }`
for object and map - `{ key: Type, value: Type/Schema }`
so this mean for all collections to support any nested type or nested schemas

File: lib/prototypes/abstract.js (hunk 8 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:16]: Changed to an array

File: lib/prototypes/abstract.js (hunk 9 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:17]: Changed to an array

**Review by tshemsedinov — APPROVED** [review]
I'd prefer to remove comments but it's ok to do it in next PR after naming improving.

File: lib/model.js [hunk 15]
```diff
@@ -1,39 +1,34 @@
...
+        continue;
       }
       const schema = new Schema(name, entity, [this]);
       this.entities.set(name, schema);
     }
+    for (const [name, entity] of last) {
+      const schema = new Schema(name, entity, [this]);
+      this.entities.set(name, schema);
+    }
+    last.clear();
```
Comment (lundibundi) [inline lib/model.js:28]: Nit:
```suggestion
```

File: lib/model.js [hunk 16]
```diff
@@ -52,7 +47,8 @@ class Model {
   reorderEntity(name, base) {
     const entity = this.entities.get(name);
     if (!entity) return;
-    for (const ref of entity.references) {
+    const { references } = entity;
+    for (const ref of references) {
```
Comment (lundibundi) [inline lib/model.js:51]: Nit: destructiring is redundant here.

File: lib/model.js [hunk 17]
```diff
@@ -1,39 +1,33 @@
...
+const { prepareTypes } = require('./types.js');
 
 class Model {
   constructor(types, entities, database = null) {
-    this.types = types;
+    this.types = prepareTypes(types);
     this.entities = new Map();
     this.database = database;
     this.order = new Set();
+    this.last = new Set();
```
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/model.js]: Done

File: lib/preprocessor.js [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+    // scalar or refernce shorthand
+    if (srcType === 'string') {
+      return this.single({ type: source });
+    }
+    if (srcType === 'object') {
+      const first = firstKey(source);
+      // schema with kind
+      if (isFirstUpper(first)) {
+        const { kinds, root } = this;
+        const [meta, fields] = omit(first, source);
```
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/preprocessor.js]: Done

File: lib/schema.js [hunk 19]
```diff
@@ -1,214 +1,111 @@
...
+    const { index, primary, unique, many } = defs;
+    const isIndex = Array.isArray(index || primary || unique);
+    if (isIndex || many) this.indexes[key] = defs;
     return isIndex;
   }
 
-  static from(raw, namespaces) {
-    return new Schema('', raw, namespaces);
+  updateMetadata(meta) {
+    this.references = new Set([...this.references, ...meta.references]);
```
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/schema.js]: Done

File: lib/schema.js (hunk 13 above)
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/schema.js]: Done

File: lib/schema.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/schema.js]: Done

File: lib/schema.js [hunk 20]
```diff
@@ -1,214 +1,111 @@
...
-          { required },
-          entry.schema ? entry : {},
-          { type: 'schema', schema }
-        );
-        this.references = new Set([...this.references, ...schema.references]);
-        this.relations = new Set([...this.relations, ...schema.relations]);
+      if (this.preprocessIndex(key, entry)) continue;
+      const child = prep.parse(entry);
+      if (child.skip) {
+        this.fields[key] = child.skip;
```
[the remark this answers is not in the document]
Comment (georgolden) [reply lib/schema.js]: Done

File: lib/schema.js (hunk 4 above)
Comment (georgolden) [reply lib/schema.js]: Done

File: lib/schema.js (hunk 5 above)
Comment (georgolden) [reply lib/schema.js]: Can be done after eslint config update

File: lib/util.js (hunk 6 above)
Comment (georgolden) [reply lib/util.js]: Done

File: lib/schema.js (hunk 7 above)
Comment (georgolden) [reply lib/schema.js:56]: Answered

File: lib/prototypes/abstract.js (hunk 8 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:17]: Impossible to use array, because field may not use `subCheck` example: `{ type: 'string' }` may not use length at all. `{ type: 'string', length: 10 }` may use length `subCheck`

Comment (georgolden) [reply lib/prototypes/abstract.js:17]: Answered

File: lib/prototypes/abstract.js (hunk 9 above)
Comment (georgolden) [reply lib/prototypes/abstract.js]: Here we need to check if there is subCheck rule in type definition. Can not be changed.

Comment (georgolden) [reply lib/prototypes/abstract.js]: Answered

File: lib/prototypes/abstract.js (hunk 10 above)
Comment (georgolden) [reply lib/prototypes/abstract.js:45]: Answered

File: lib/prototypes/collections.js (hunk 11 above)
Comment (georgolden) [reply lib/prototypes/collections.js]: Done

File: lib/prototypes/collections.js (hunk 12 above)
Comment (georgolden) [reply lib/prototypes/collections.js:92]: Answered

File: lib/model.js (hunk 16 above)
Comment (georgolden) [reply lib/model.js]: Done

**Comment by tshemsedinov** [comment]
Landed in https://github.com/metarhia/metaschema/commit/ee1bca5e7512ede93b0dfdc9d534eef0f03d14a2
